// serverless-esbuild/dist/utils.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.trimExtension = exports.zip = exports.humanSize = exports.findProjectRoot = exports.findUp = exports.spawnProcess = exports.SpawnError = void 0;
const bestzip_1 = require("bestzip");
const archiver = require("archiver");
const childProcess = require("child_process");
const fs = require("fs-extra");
const path = require("path");
const os = require("os");
const ramda_1 = require("ramda");
class SpawnError extends Error {
    constructor(message, stdout, stderr) {
        super(message);
        this.stdout = stdout;
        this.stderr = stderr;
    }
    toString() {
        return `${this.message}\n${this.stderr}`;
    }
}
exports.SpawnError = SpawnError;
/**
 * Executes a child process without limitations on stdout and stderr.
 * On error (exit code is not 0), it rejects with a SpawnProcessError that contains the stdout and stderr streams,
 * on success it returns the streams in an object.
 * @param {string} command - Command
 * @param {string[]} [args] - Arguments
 * @param {Object} [options] - Options for child_process.spawn
 */
function spawnProcess(command, args, options) {
    return new Promise((resolve, reject) => {
        const child = childProcess.spawn(command, args, options);
        let stdout = '';
        let stderr = '';
        // Configure stream encodings
        child.stdout.setEncoding('utf8');
        child.stderr.setEncoding('utf8');
        // Listen to stream events
        child.stdout.on('data', (data) => {
            stdout += data;
        });
        child.stderr.on('data', (data) => {
            stderr += data;
        });
        child.on('error', (err) => {
            reject(err);
        });
        child.on('close', (exitCode) => {
            if (exitCode !== 0) {
                reject(new SpawnError(`${command} ${(0, ramda_1.join)(' ', args)} failed with code ${exitCode}`, stdout, stderr));
            }
            else {
                resolve({ stdout, stderr });
            }
        });
    });
}
exports.spawnProcess = spawnProcess;
/**
 * Find a file by walking up parent directories
 */
function findUp(names, directory = process.cwd()) {
    const absoluteDirectory = path.resolve(directory);
    if (typeof names === 'string') {
        names = [names];
    }
    /* For vs. .forEach so it can exit when we get a hit. */
    for (const name of names) {
        if (fs.existsSync(path.join(directory, name))) {
            return directory;
        }
    }
    const { root } = path.parse(absoluteDirectory);
    if (absoluteDirectory === root) {
        return undefined;
    }
    return findUp(names, path.dirname(absoluteDirectory));
}
exports.findUp = findUp;
/**
 * Forwards `rootDir` or finds project root folder.
 */
function findProjectRoot(rootDir) {
    return rootDir !== null && rootDir !== void 0 ? rootDir : findUp(['yarn.lock', 'package-lock.json']);
}
exports.findProjectRoot = findProjectRoot;
const humanSize = (size) => {
    const i = Math.floor(Math.log(size) / Math.log(1024));
    const sanitized = (size / Math.pow(1024, i)).toFixed(2);
    return `${sanitized} ${['B', 'KB', 'MB', 'GB', 'TB'][i]}`;
};
exports.humanSize = humanSize;
const zip = (zipPath, filesPathList, useNativeZip = false) => __awaiter(void 0, void 0, void 0, function* () {
    // create a temporary directory to hold the final zip structure
    const tempDirName = `${path.basename(zipPath).slice(0, -4)}-${Date.now().toString()}`;
    const tempDirPath = path.join(os.tmpdir(), tempDirName);
    fs.mkdirpSync(tempDirPath);
    // copy all required files from origin path to (sometimes modified) target path
    yield Promise.all(filesPathList.map((file) => fs.copy(file.rootPath, path.join(tempDirPath, file.localPath))));
    // prepare zip folder
    fs.mkdirpSync(path.dirname(zipPath));
    if (useNativeZip) {
        // zip the temporary directory
        yield (0, bestzip_1.bestzip)({
            source: '*',
            destination: zipPath,
            cwd: tempDirPath,
        });
        // delete the temporary folder
        fs.rmdirSync(tempDirPath, { recursive: true });
    }
    else {
        const zip = archiver.create('zip');
        const output = fs.createWriteStream(zipPath);
        // write zip
        output.on('open', () => {
            zip.pipe(output);
            filesPathList.forEach((file) => {
                const stats = fs.statSync(file.rootPath);
                if (stats.isDirectory())
                    return;
                zip.append(fs.readFileSync(file.rootPath), {
                    name: file.localPath,
                    mode: stats.mode,
                    date: new Date(0), // necessary to get the same hash when zipping the same content
                });
            });
            zip.finalize();
        });
        return new Promise((resolve, reject) => {
            output.on('close', () => {
                // delete the temporary folder
                fs.rmdirSync(tempDirPath, { recursive: true });
                resolve();
            });
            zip.on('error', (err) => reject(err));
        });
    }
});
exports.zip = zip;
function trimExtension(entry) {
    return entry.slice(0, -path.extname(entry).length);
}
exports.trimExtension = trimExtension;
